designs/disruption-termination-grace-period.md

```diff
@@ -0,0 +1,122 @@
+# Disruption Termination Grace Period
+
+## Motivation
+Users are requesting the ability to control how long karpenter will wait for the deletion of an individual node to complete before forcefully terminating the node regardless of pod status. [#743](https://github.com/kubernetes-sigs/karpenter/issues/743). This supports two primary use cases.
+* Cluster admins who want to ensure that nodes are cycled after a given period of time, regardless of user-defined disruption controls (such as PDBs or preStop hooks) that might prevent eviction of a pod beyond the configured limit. This could be to satisfy security requirements or for convenience.
+* Cluster admins who want to allow users to protect long-running jobs from being interrupted by node disruption, up to a configured limit.
+
+This design piggybacks off the existing disruption-controls design, adding in a new `terminationGracePeriod` field to the `v1beta/NodePool` `Disruption` spec. This field seems intuitively related to the other disruption controls that exist within the disruption block.
+
+## Proposed Spec
+
+```yaml
+apiVersion: karpenter.sh/v1beta1
+kind: NodePool
+metadata:
+  name: default
+spec: # This is not a complete NodePool Spec.
+  disruption:
+    consolidationPolicy: WhenUnderutilized || WhenEmpty
+    consolidateAfter: 10m || Never # metav1.Duration
+    expireAfter: 10m || Never # Equivalent to v1alpha5 TTLSecondsUntilExpired
+    terminationGracePeriod: 24h || nil # metav1.Duration
```

Comment:
Since we already are scoped to disruption, I'd suggest dropping the termination bit. We typically try to design the API such that you can speak the dotted expansion of the config, E.g.

disruption.gracePeriod is the "node pool's disruption grace period" 

disruption.terminationGracePeriod is nonsensical to speak, and intermingled concepts of disruption and termination.

Reply:
The disruption field isn't limited to node termination behavior. I think it's reasonable and leaves no room for misunderstanding to keep the full name `disruption.terminationGracePeriod`.

Reply:
Disruption encompasses both the disruption and termination controllers, which a disruptionGracePeriod in the context of the disruption controller doesn't really make much sense. Prefacing with `termination` makes it more explicit in what process it's referencing. WDYT @ellistarn ?

Reply:
Makes me wonder if it should even be part of disruption, then? What other fields in disruption control termination?

Reply:
> Makes me wonder if it should even be part of disruption, then

Agree with @ellistarn. Is this just a property of the NodeClaim? Can we imagine that if you initiate a deletion request on a NodeClaim as a standalone item that it just deprovisions on its own? I'd argue, yes. I wonder if (similar to pod) this fits best in `spec.template.spec.terminationGracePeriod`?

Reply:
Whether or not it's a NodeClaim property depends on the behavior we'd like to see. If you want to be able to lower the `terminationGracePeriod` in order to force deletion of live nodes, you'd want it to not be a NodeClaim field, since that wouldn't be changed if the NodePool's `terminationGracePeriod` was updated. Are NodeClaim's mutuble? Would a cluster admin be able to modify the NodeClaim for a Node in order to force early termination of a Node if desired? Maybe it's sufficient for the cluster admin to have to perform a force drain on a terminating Node if they want to force early termination.

In my company, we run 24h `terminationGracePeriod`'s in production, and want the ability to forcefully tear down the Node earlier if desired. Seems like that would still be possible if it's a field on the NodeClaim via `kubectl drain`, but I just wanted to lay out the implications of one choice over another.

Reply:
> Are NodeClaim's mutuble

Reply:
> Are NodeClaim's mutuble

Yes, NodeClaims are mutable; though, mutating certain fields on them may not have the behavior that's expected today since mutating some things like the labels/annotations won't necessarily have any affect on Node that they own.

 > you'd want it to not be a NodeClaim field, since that wouldn't be changed if the NodePool's terminationGracePeriod was updated

One option here is to let this field be driftable, if a NodePool's value for this changes, then all NodeClaims that are owned by that NodePool are drifted.

Reply:
I kinda dislike that updating a NodePool could both have and not have large implications for all the NodeClaims that belong to it. 😅 Might be worth trying to aim for consistency of behavior there to avoid surprises for Karpenter users.

Even if the field is immutable on the NodeClaim, I think a viable path exists for a cluster-admin to force termination of the Node via `kubectl drain`. It's the same path that already exists for Karpenter users today too. But I'm interested in hearing what others think about what the expectations should be here.
1. Updating NodePool's `terminationGracePeriod` immediately changes the timeout for all existing Nodes that belong to the NodePool, which could result in mass termination of Nodes (possibly accidentally). Extension question, can this termination bypass the `nodepool.spec.disruption.budgets.nodes` value configured for the NodePool?
2. Updating the NodePool's `terminationGracePeriod` only changes the `terminationGracePeriod` for subsequently created NodeClaims and Nodes.

Reply:
IMO, if we decided to add this field directly on the NodeClaim, having one field in the NodePool that behaves differently in the `template` block than the other fields in the `template` block is pretty unexpected. Today, if you make a change to the `template` block, `NodeClaims` get marked drifted and you will get nodes rolling to bring your Nodes into the new desired state that you set in the `NodePool`. We are communicating a relationship similar to the Pod/Deployment objects with the NodeClaim/NodePool through the `template` field. If you update the `terminationGracePeriodSeconds` of the pod in the deployment object, you get a pod roll, I'd expect similar behavior here.

Reply:
> So, the other line that we are probably drawing here is:
> 
> 1. If we want to be able to make a change to `terminationGracePeriod` that automatically affects all NodeClaims, that field needs to live in the NodePool, outside of the `template` block

That fits, _except_ that we might want to allow changes to the `.spec.template.metadata` of a NodePool, which triggers a metadata change via updates to the NodeClaims that it owns. For example, adding an annotation `whatever.fictional.example/soft-node-drain-after: 2024-01-01T00:00:00+0000` (which no cluster operator would thank you for).

> 2. If we are alright having a rolling update for `terminationGracePeriod` if that value is changed, then I think it's fine in the `template` block

Sure.

---

We should avoid having any annotations or labels that mutate a field that is in the `spec` of a NodeClaim. So 
`soft-node-drain-after` is fine (because it doesn't override `terminationGracePeriod` or any other field), but `terminationGracePeriodAdjustment` would be a poor choice.

Reply:
Does modifying a pod's terminationGracePeriodSeconds in a deployment's pod template spec result in a rollout? If not, I could see the case for following suit and special casing drift.

Reply:
Yes, it does. Just double checked that modifying a pod's terminationGracePeriodSeconds from 30 -> 60 in the Deployment.spec.template.spec did roll the pods.

```
$ kubectl edit deploy -n default
deployment.apps/hello-world-nginx edited

$ kubectl get pod -n default    
NAME                                 READY   STATUS              RESTARTS   AGE
hello-world-nginx-5fd5b65b9d-zpksw   1/1     Running             0          32h
hello-world-nginx-84556d7755-cgzzf   1/1     Running             0          3s
hello-world-nginx-84556d7755-d6sbh   0/1     ContainerCreating   0          1s
```

Reply:
It's not unreasonable to follow suit from a principle of least surprise perspective. Perhaps nodes of a given spec may need different amounts of time to terminate. I could imagine cluster admins being annoyed at this behavior in the rare case it needs to be changed. I do like the fact it applies to the Node claim, rather than the node pool, as a standalone nodeclaim could have similar drain requirements.

Reply:
> I could imagine cluster admins being annoyed at this behavior in the rare case it needs to be changed

The more I think about this, the more I'm convinced that it makes sense just to one-off this as something that can be updated in-place. I can see updating the value causing a rollout across all nodes as extremely annoying and the value that it gives us is pretty minimal.

Realistically, I'm failing to think of strong sets of use-cases where a user would want to rollout a change to terminationGracePeriod that needs to be closely tied to other changes that they are making to the NodePool at the same time.